src/ui/bibliography-renderer/render-list-content-script.ts

```diff
@@ -0,0 +1,74 @@
+import { Reference } from "../../model/reference.model";
+
+declare const webviewApi: any;
+
+export default function(_context) { 
+	return {
+		plugin: function(markdownIt, _options) {
+			const contentScriptId = _context.contentScriptId;
+
+			/* Appends a new custom token for references list */
+			markdownIt.core.ruler.push("reference_list", async state => {
+
+				/* Collect references from the note body using Depth-first-search */
+				const ids: Reference[] = [];
+				dfs(state.tokens);
+				
+				function dfs (children: any[]): void {
+					if (!children) return;
+
+					/* Search for three consecutive tokens: "link_open", "text", and "link_close" */
+					for (let i = 1; i < children.length - 1; i++) {
+						const curr = children[i], prev = children[i-1], next = children[i+1];
+						if (
+							prev["type"] === "link_open" &&
+							curr["type"] === "text" &&
+							next["type"] === "link_close" &&
+							curr.content && curr.content.length > 1 &&
+							curr.content.startsWith("@")
+						) {
+							const id = curr.content.substring(1);
+							ids.push(id);
+						} else {
+							if (curr["children"]) dfs(curr["children"]);
+						}
+					}
+					// first and last child that were not traversed previously
+					const last = children[children.length - 1], first = children[0];
+					if (last["children"]) dfs(last["children"]);
+					if (first["children"]) dfs(first["children"]);
+				};
+				
+				/* Append reference_list token */
+				let token = new state.Token("reference_list", "", 0);
+				console.log(ids);
+				token.attrSet("refs", ids);
+				state.tokens.push(token);
+			});
+			
+			/* Define how to render the previously defined token */
+			markdownIt.renderer.rules["reference_list"] = renderReferenceList;
+
+			function renderReferenceList (tokens, idx, options) {
+				let IDs: string[] = tokens[idx]["attrs"][0][1];
+				if (IDs.length === 0) return "";
+
+				const script: string = `
+					webviewApi.postMessage("${contentScriptId}", ${JSON.stringify(IDs)}).then(refs => {
+						const referenceListView = document.getElementById("references_list");
+						const referenceTitleView = document.getElementById("references_title");
+						if (refs.length > 0) referenceTitleView.style.display = "block";
+						refs.forEach(ref => referenceListView.innerHTML += ref);
+					});
+					return false;
+				`;
+
+				return `
+					<h1 id="references_title" style="display:none">References</h1>
+					<ul id="references_list"></ul>
+					<style onload='${script.replace(/\n/g, ' ')}'/>
```

Comment:
Any reason you used a style tag for this? Wouldn't it be possible to wrap your script string simply in a `<script>` tag? I think that would be more proper and perhaps more reliable too.

Reply:
> Wouldn't it be possible to wrap your script string simply in a <script> tag? I think that would be more proper and perhaps more reliable too.

I tried that approach as it seems to be the most natural but for, some reason, it does not seem to work.

Reply:
Ok, I think this is fragile code that may break at some point or even work in some operating systems but not others. I think it didn't work with the script tag because you need to wait till the "webviewApi" is available. For this you can use the same pattern - just keep checking `typeof webviewApi` until it becomes available, then run your initialisation code.

Reply:
Whatever I put in the script tag does not get executed.

Reply:
I'm wondering if you're looking at all this the wrong way around. The plugin sets the HTML, and has all the references, so instead of adding a script that post messages, can't you directly set the HTML to the correct value from the plugin? Then you won't need a script at all inside the content script.

Reply:
In the documentation, there is no description of such approach (in markdown-it plugin section)

Reply:
It's not something about documentation, since it's specific to your plugin. But come to think of it, what I'm suggesting wouldn't work since the reference data is in the plugin process, and the content script in the main process, so there would still be a need for some kind of ipc call. So let's leave it as it is for now.